File: slate/slasm/function.py
from typing import Dict, Iterator, List, Optional, Set, Tuple

from slate.slasm.instruction import Instruction, OpCode

class BasicBlock:
    __TERMINATING_INSTRUCTIONS = set([
        OpCode.JUMP,
        OpCode.COND_JUMP,
        OpCode.RET
    ])

    def __init__(self) -> None:
        self.__instructions : List[Instruction] = []

    def append_instr(self, instr: Instruction) -> None:
        if self.is_terminated():
            raise Exception("Cannot append an instruction to a terminated block!")

        self.__instructions.append(instr)

    def is_terminated(self) -> bool:
        return len(self.__instructions) != 0 and self.__instructions[-1].opcode in BasicBlock.__TERMINATING_INSTRUCTIONS

    def __iter__(self) -> Iterator[Instruction]:
        return self.__instructions.__iter__()

    def __getitem__(self, idx: int) -> Instruction:
        return self.__instructions[idx]

class Function:
    def __init__(self, name: str, params: Set[str], locals: Set[str], returns_value: bool) -> None:
        self.__name = name
        self.__params = set(params)
        self.__locals = set(locals)
        self.__returns_value = returns_value
        self.__basic_blocks : Dict[str, BasicBlock] = {}
        self.__entry : Optional[str] = None

    def add_basic_block(self, name: str, basic_block: BasicBlock) -> None:
        if name in self.__basic_blocks:
            raise Exception(f"Basic block with name {name} already exists!")

        self.__basic_blocks[name] = basic_block

    def contains_nonterminated_basic_block(self) -> bool:
        return not all([bb.is_terminated() for bb in self.__basic_blocks.values()])

    @property
    def name(self) -> str:
        return self.__name

    @property
    def num_params(self) -> int:
        return len(self.__params)

    @property
    def num_locals(self) -> int:
        return len(self.__locals)

    @property
    def returns_value(self) -> bool:
        return self.__returns_value

    @property
    def entry(self) -> str:
        if self.__entry is None:
            raise Exception("Function's entry has not been set!")

        return self.__entry

    @entry.setter
    def entry(self, basic_block_name: str) -> None:
        if basic_block_name not in self.__basic_blocks:
            raise Exception(f"Basic block with name {basic_block_name} does not exist!")

        self.__entry = basic_block_name

    @property
    def basic_blocks(self) -> List[Tuple[str, BasicBlock]]:
        return list(self.__basic_blocks.items())

    @property
    def params(self) -> Set[str]:
        return set(self.__params)

    @property
    def locals(self) -> Set[str]:
        return set(self.__locals)

File: src/slasm/function.rs
use crate::slasm::basic_block::BasicBlock;
use std::collections::HashMap;

pub struct Function {
    name: String,
    params: HashMap<String, u64>,
    locals: HashMap<String, u64>,
    returns: Vec<u64>,
    basic_blocks: HashMap<String, BasicBlock>,
    entry: Option<String>,
}

impl Function {
    pub fn new(
        name: String,
        params: HashMap<String, u64>,
        locals: HashMap<String, u64>,
        returns: Vec<u64>,
    ) -> Function {
        assert!(returns.iter().all(|ret| ret > &0), "Functions cannot return 0-sized values!");

        Function {
            name: name,
            params: params,
            locals: locals,
            returns: returns,
            basic_blocks: HashMap::new(),
            entry: None,
        }
    }

    pub fn add_basic_block(&mut self, name: String, basic_block: BasicBlock) {
        assert!(
            !self.basic_blocks.contains_key(&name),
            "Function '{}' already contains a basic block with name '{}'!",
            self.name,
            name
        );

        self.basic_blocks.insert(name, basic_block);
    }

    pub fn contains_nonterminated_basic_block(&self) -> bool {
        !self.basic_blocks.values().all(|bb| bb.is_terminated())
    }

    pub fn name(&self) -> &String {
        &self.name
    }

    pub fn params(&self) -> &HashMap<String, u64> {
        &self.params
    }

    pub fn locals(&self) -> &HashMap<String, u64> {
        &self.locals
    }

    pub fn returns(&self) -> &Vec<u64> {
        &self.returns
    }

    pub fn total_returns_size(&self) -> u64 {
        self.returns.iter().sum()
    }

    pub fn basic_blocks(&self) -> &HashMap<String, BasicBlock> {
        &self.basic_blocks
    }

    pub fn is_procedure(&self) -> bool {
        self.total_returns_size() == 0
    }

    pub fn entry(&self) -> &String {
        if let Some(name) = &self.entry {
            return name;
        }

        panic!("Function '{}' does not have a set entry!", self.name);
    }

    pub fn set_entry(&mut self, value: String) {
        assert!(
            self.basic_blocks.contains_key(&value),
            "Function '{}' does not contain a basic block with name '{}'!",
            self.name,
            value
        );

        self.entry = Some(value);
    }
}
